driver/csiplugin/controllerserver.go

```diff
@@ -400,6 +413,31 @@ func (cs *ScaleControllerServer) CreateVolume(ctx context.Context, req *csi.Crea
 	scaleVol.PrimaryFSMount = PFSMount
 	scaleVol.PrimarySLnkPath = PSLnkPath
 
+	volSrc := req.GetVolumeContentSource()
+	isSnapSource := false
+	snapIdMembers := scaleSnapId{}
+	if volSrc != nil {
+		if volSrc.GetVolume() != nil {
+			return nil, status.Error(codes.Unimplemented, "Volume as volume content source is not supported")
+		}
+
+		srcSnap := volSrc.GetSnapshot()
+		if srcSnap != nil {
+			snapId := srcSnap.GetSnapshotId()
+			snapIdMembers, err = cs.GetSnapIdMembers(snapId)
+			if err != nil {
+				glog.Errorf("CreateVolume [%s]: Invalid snapshot ID %s [%v]", volName, snapId, err)
+				return nil, err
+			}
+			err = cs.ValidateSnapId(&snapIdMembers, scaleVol, PCid)
```

Comment:
In case of volume create from snapshot there are duplicate check being added because of ValidateSnapId function, some the check are there after this function.

Reply:
Moved validateSnapId call below other validations and optimized some validations in the function to avoid repeated checks.